### PR TITLE
Support to configure multiple Github Organizations

### LINKED ISSUE #552: Enhancement: Make it possible to configure multiple GitHub Organizations
Hi,

The existing configuration allows you to lock the plugin to a single organization.
This is fine, but sometimes you might have an OSS community that manages a few projects under different organizations. 
In that situation you would not be able to subscribe to issues from any other organization other than the configured one in the System Console.

I would like to propose to allow the usage of multiple GitHub organizations (i.e. separated by space or comma)
This would allow users to subscribe for notifications only from the mentioned organizations and the repositories inside them.

### PR DESCRIPTION
**Pull request by khareyash05**
#### Summary

Configure support to multiple Github organizations 
#### Ticket Link
Fixes #552

### CODE REVIEW

**Comment by khareyash05** [comment]
hi @hanzei . I am actually facing a problem in [https://github.com/mattermost/mattermost-plugin-github/blob/master/server/plugin/plugin.go#L166](https://github.com/mattermost/mattermost-plugin-github/blob/master/server/plugin/plugin.go#L166) . It expect a single organization name for creating a new Client for graphql. How can this be tackled for multiple orgs?

**Comment by hanzei** [comment]
Does it make sense to move the org name out of the graphql client and pass the list of orgs as part of  `GetLHSData`?

**Comment by khareyash05** [comment]
Will try and let you know! That might help Thanks

**Comment by khareyash05** [comment]
> Does it make sense to move the org name out of the graphql client and pass the list of orgs as part of `GetLHSData`?

@hanzei If i understood it correctly, graphql.NewClient expects an organization. How can we tackle that issue? Passing organizations through `GetLHSData` will work for utility function `getLHSData` but how will be tackle the NewClient function? Can we consider passing orgs as concatenated string? Through my limited understanding of Go, this was I could think of. Kindly help

**Comment by hanzei** [comment]
Sorry for the delay.

If we are passing the org as part of `GetLHSData`, we can remove `org` from `graphql.Client`. It's no longer needed.

File: server/plugin/configuration.go [hunk 1]
```diff
@@ -25,19 +25,19 @@ import (
...
-	EnableLeftSidebar           bool   `json:"enableleftsidebar"`
-	EnablePrivateRepo           bool   `json:"enableprivaterepo"`
-	ConnectToPrivateByDefault   bool   `json:"connecttoprivatebydefault"`
-	EncryptionKey               string `json:"encryptionkey"`
-	EnterpriseBaseURL           string `json:"enterprisebaseurl"`
-	EnterpriseUploadURL         string `json:"enterpriseuploadurl"`
-	EnableCodePreview           string `json:"enablecodepreview"`
-	EnableWebhookEventLogging   bool   `json:"enablewebhookeventlogging"`
-	UsePreregisteredApplication bool   `json:"usepreregisteredapplication"`
+	GitHubOrgs                  []string `json:"githuborgs"`
```
Comment (hanzei) [inline server/plugin/configuration.go:28]: How does this change affect existing installation? E.g., does the previous value from `githuborg` get carried over to `githuborgs`?

File: server/plugin/plugin.go [hunk 2]
```diff
@@ -163,7 +163,8 @@ func (p *Plugin) githubConnectUser(ctx context.Context, info *GitHubUserInfo) *g
 
 func (p *Plugin) graphQLConnect(info *GitHubUserInfo) *graphql.Client {
 	conf := p.getConfiguration()
-	return graphql.NewClient(p.client.Log, *info.Token, info.GitHubUsername, conf.GitHubOrg, conf.EnterpriseBaseURL)
+	// an issue here to check of graphql client
```
Comment (hanzei) [inline server/plugin/plugin.go:166]: What does that mean?

File: server/plugin/plugin.go [hunk 3]
```diff
@@ -953,32 +954,39 @@ func (p *Plugin) HasUnreads(info *GitHubUserInfo) bool {
 func (p *Plugin) checkOrg(org string) error {
 	config := p.getConfiguration()
 
-	configOrg := strings.TrimSpace(config.GitHubOrg)
-	if configOrg != "" && configOrg != org && strings.ToLower(configOrg) != org {
-		return errors.Errorf("only repositories in the %v organization are supported", configOrg)
+	for _, org := range config.GitHubOrgs {
+		configOrg := strings.TrimSpace(org)
```
Comment (hanzei) [inline server/plugin/plugin.go:958]: Isn't trimming the spaces already done as part of `sanitize()`?

File: server/plugin/plugin.go [hunk 4]
```diff
@@ -953,32 +954,39 @@ func (p *Plugin) HasUnreads(info *GitHubUserInfo) bool {
...
-func (p *Plugin) isUserOrganizationMember(githubClient *github.Client, user *github.User, organization string) bool {
-	if organization == "" {
-		return false
-	}
+func (p *Plugin) isUserOrganizationMember(githubClient *github.Client, user *github.User, organizations []string) bool {
+	var isMember bool
+	for _, org := range organizations {
+		if org == "" {
+			return false
+		}
```
Comment (hanzei) [inline server/plugin/plugin.go:971]: Can this case happen?

File: server/plugin/plugin.go [hunk 5]
```diff
@@ -953,32 +954,39 @@ func (p *Plugin) HasUnreads(info *GitHubUserInfo) bool {
...
-	isMember, _, err := githubClient.Organizations.IsMember(context.Background(), organization, *user.Login)
-	if err != nil {
-		p.client.Log.Warn("Failled to check if user is org member", "GitHub username", *user.Login, "error", err.Error())
-		return false
+		isMember, _, err := githubClient.Organizations.IsMember(context.Background(), org, *user.Login)
+		if err != nil {
+			p.client.Log.Warn("Failled to check if user is org member", "GitHub username", *user.Login, "error", err.Error())
+			return false
+		}
+		isMember = isMember || true
```
Comment (hanzei) [inline server/plugin/plugin.go:978]: That seems off.

File: server/plugin/plugin.go [hunk 6]
```diff
@@ -953,32 +954,39 @@ func (p *Plugin) HasUnreads(info *GitHubUserInfo) bool {
...
 }
 
 func (p *Plugin) isOrganizationLocked() bool {
 	config := p.getConfiguration()
-	configOrg := strings.TrimSpace(config.GitHubOrg)
-
+	configOrg := ""
+	for _, org := range config.GitHubOrgs {
+		configOrg = strings.TrimSpace(org)
+	}
```
Comment (hanzei) [inline server/plugin/plugin.go:989]: Is it enough to check for `len(config.GitHubOrgs) > 0`?

File: server/plugin/telemetry.go [hunk 7]
```diff
@@ -39,7 +39,7 @@ func (p *Plugin) SendDailyTelemetry() {
 		"connected_user_count":          connectedUserCount,
 		"is_oauth_configured":           config.IsOAuthConfigured(),
 		"is_sass":                       config.IsSASS(),
-		"is_organization_locked":        config.GitHubOrg != "",
+		"is_organization_locked":        config.GitHubOrgs,
```
Comment (hanzei) [inline server/plugin/telemetry.go:42]: The values of `is_organization_locked` is still a `bool`. Should we use `p.isOrganizationLocked()` instead?

File: server/plugin/utils.go [hunk 8]
```diff
@@ -18,36 +18,36 @@ import (
...
 
 	return fmt.Sprintf(query, orgField, searchTerm)
 }
 
-func buildSearchQuery(query, username, org string) string {
+func buildSearchQuery(query string, username string, orgs []string) string {
 	orgField := ""
-	if len(org) != 0 {
-		orgField = fmt.Sprintf("org:%v", org)
+	if len(orgs) != 0 {
```
Comment (hanzei) [inline server/plugin/utils.go:49]:
```suggestion
	if len(orgs) > 0 {
```

File: server/plugin/command.go [hunk 9]
```diff
@@ -682,15 +682,15 @@ func getAutocompleteData(config *Configuration) *model.AutocompleteData {
 	subscriptionsAdd.AddTextArgument("Owner/repo to subscribe to", "[owner/repo]", "")
 	subscriptionsAdd.AddNamedTextArgument("features", "Comma-delimited list of one or more of: issues, pulls, pulls_merged, pushes, creates, deletes, issue_creations, issue_comments, pull_reviews, label:\"<labelname>\". Defaults to pulls,issues,creates,deletes", "", `/[^,-\s]+(,[^,-\s]+)*/`, false)
 
-	if config.GitHubOrg != "" {
-		subscriptionsAdd.AddNamedStaticListArgument("exclude-org-member", "Events triggered by organization members will not be delivered (the organization config should be set, otherwise this flag has not effect)", false, []model.AutocompleteListItem{
+	for _, org := range config.GitHubOrgs {
+		subscriptionsAdd.AddNamedStaticListArgument("exclude-org-member-"+org, "Events triggered by organization members of "+org+" will not be delivered (the organization config should be set, otherwise this flag has no effect)", false, []model.AutocompleteListItem{
```
Comment (hanzei) [inline server/plugin/command.go:686]: Let's keep that flag as it is for now. The gist of the flag stays the same even with multiple orgs: It limits the subscription to all configured orgs.

File: server/plugin/api.go [hunk 10]
```diff
@@ -565,15 +565,15 @@ func (p *Plugin) getConnected(c *Context, w http.ResponseWriter, r *http.Request
 		GitHubUsername      string                 `json:"github_username"`
 		GitHubClientID      string                 `json:"github_client_id"`
 		EnterpriseBaseURL   string                 `json:"enterprise_base_url,omitempty"`
-		Organization        string                 `json:"organization"`
+		Organizations       []string               `json:"organization"`
```
Comment (hanzei) [inline server/plugin/api.go:568]: Does the webapp need to get updated because of this change?

File: server/plugin/api.go [hunk 11]
```diff
@@ -1226,34 +1226,36 @@ func getRepositoryListByOrg(c context.Context, org string, githubClient *github.
...
-			return
-		}
-	} else {
-		allRepos, statusCode, err = getRepositoryListByOrg(c.Ctx, org, githubClient, opt)
-		if err != nil {
-			if statusCode == http.StatusNotFound {
-				allRepos, err = getRepositoryList(c.Ctx, org, githubClient, opt)
-				if err != nil {
+	for _, org := range orgs {
+		if org == "" {
```
Comment (hanzei) [inline server/plugin/api.go:1237]: Can this case happen?

**Review by hanzei — CHANGES_REQUESTED** [review]
Thanks for the PR @khareyash05 :+1: 

I left my comments down below. To summarize, I'm not 100% sure that the changes work. Did you test them?

**Comment by mickmister** [comment]
Hi @khareyash05, checking in to see how you're doing. Do you have any questions about this task?

Thanks @khareyash05!!

**Comment by khareyash05** [comment]
> Hi @khareyash05, checking in to see how you're doing. Do you have any questions about this task?
> 
> Thanks @khareyash05!!

Hey @mickmister Not able to understand the work to do. Closing the PR. Thanks.

**Pull request closed without merging**